File: sdks/android-java/src/main/java/dev/nucleus/network/ApiClient.java
package dev.nucleus.network;

import org.json.JSONObject;
import org.json.JSONArray;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import dev.nucleus.NucleusCallback;
import dev.nucleus.NucleusException;
import dev.nucleus.NucleusUser;
import dev.nucleus.model.*;
import okhttp3.*;

public class ApiClient {
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    private final OkHttpClient client;
    private final String baseUrl;
    private final String publishableKey;
    private String accessToken;
    private final Executor executor = Executors.newCachedThreadPool();

    public ApiClient(String publishableKey, String baseUrl) {
        this.publishableKey = publishableKey;
        this.baseUrl = baseUrl.replaceAll("/$", "");
        this.client = new OkHttpClient();
    }

    public void setToken(String token) { this.accessToken = token; }

    private Request.Builder requestBuilder(String path) {
        Request.Builder b = new Request.Builder()
            .url(baseUrl + path)
            .header("Content-Type", "application/json")
            .header("X-Nucleus-Publishable-Key", publishableKey);
        if (accessToken != null) b.header("Authorization", "Bearer " + accessToken);
        return b;
    }

    private JSONObject execute(Request request) throws NucleusException {
        try (Response response = client.newCall(request).execute()) {
            String body = response.body() != null ? response.body().string() : "";
            if (!response.isSuccessful()) throw new NucleusException("API error (" + response.code() + "): " + body);
            if (body.isEmpty() || response.code() == 204) return new JSONObject();
            return new JSONObject(body);
        } catch (NucleusException e) { throw e; }
        catch (Exception e) { throw new NucleusException("Network error", e); }
    }

    private JSONObject post(String path, JSONObject body) throws NucleusException {
        Request req = requestBuilder(path).post(RequestBody.create(body.toString(), JSON)).build();
        return execute(req);
    }

    private JSONObject get(String path) throws NucleusException {
        return execute(requestBuilder(path).get().build());
    }

    private JSONObject put(String path, JSONObject body) throws NucleusException {
        return execute(requestBuilder(path).put(RequestBody.create(body.toString(), JSON)).build());
    }

    private JSONObject patch(String path, JSONObject body) throws NucleusException {
        return execute(requestBuilder(path).patch(RequestBody.create(body.toString(), JSON)).build());
    }

    private void delete(String path) throws NucleusException {
        execute(requestBuilder(path).delete().build());
    }

    // --- Async wrapper ---
    private <T> void async(NucleusCallback<T> cb, AsyncTask<T> task) {
        executor.execute(() -> {
            try { T result = task.run(); cb.onSuccess(result); }
            catch (NucleusException e) { cb.onError(e); }
            catch (org.json.JSONException e) { cb.onError(new NucleusException("JSON parse error: " + e.getMessage(), e)); }
        });
    }

    @FunctionalInterface
    interface AsyncTask<T> { T run() throws NucleusException, org.json.JSONException; }

    // --- Auth ---
    public void signIn(String email, String password, NucleusCallback<AuthResult> cb) {
        async(cb, () -> {
            JSONObject body = new JSONObject().put("email", email).put("password", password);
            JSONObject json = post("/v1/auth/sign-in", body);
            return new AuthResult(NucleusUser.fromJson(json.getJSONObject("user")), NucleusSession.fromJson(json.getJSONObject("session")));
        });
    }

    public void signUp(String email, String password, String firstName, String lastName, NucleusCallback<AuthResult> cb) {
        async(cb, () -> {
            JSONObject body = new JSONObject().put("email", email).put("password", password);
            if (firstName != null) body.put("first_name", firstName);
            if (lastName != null) body.put("last_name", lastName);
            JSONObject json = post("/v1/auth/sign-up", body);
            return new AuthResult(NucleusUser.fromJson(json.getJSONObject("user")), NucleusSession.fromJson(json.getJSONObject("session")));
        });
    }

    public void signOut(NucleusCallback<Void> cb) {
        async(cb, () -> { post("/v1/auth/sign-out", new JSONObject()); return null; });
    }

    // --- OAuth ---
    public String getOAuthUrl(String provider, String redirectUri) {
        return baseUrl + "/v1/oauth/" + provider + "/authorize?redirect_uri=" + redirectUri + "&publishable_key=" + publishableKey;
    }

    public void exchangeOAuthCode(String code, String redirectUri, NucleusCallback<AuthResult> cb) {
        async(cb, () -> {
            JSONObject body = new JSONObject().put("code", code).put("redirect_uri", redirectUri);
            JSONObject json = post("/v1/oauth/token", body);
            return new AuthResult(NucleusUser.fromJson(json.getJSONObject("user")), NucleusSession.fromJson(json.getJSONObject("session")));
        });
    }

    // --- MFA ---
    public void mfaTotpSetup(NucleusCallback<MfaSetupResult> cb) {
        async(cb, () -> {
            JSONObject json = post("/v1/auth/mfa/totp/setup", new JSONObject());
            return new MfaSetupResult(json.getString("secret"), json.getString("qr_uri"));
        });
    }

    public void mfaTotpVerify(String code, NucleusCallback<Boolean> cb) {
        async(cb, () -> post("/v1/auth/mfa/totp/verify", new JSONObject().put("code", code)).getBoolean("verified"));
    }

    public void mfaSmsSend(String phone, NucleusCallback<Void> cb) {
        async(cb, () -> { post("/v1/auth/mfa/sms/send", new JSONObject().put("phone", phone)); return null; });
    }

    public void mfaSmsVerify(String code, NucleusCallback<Boolean> cb) {
        async(cb, () -> post("/v1/auth/mfa/sms/verify", new JSONObject().put("code", code)).getBoolean("verified"));
    }

    // --- User ---
    public void getUser(NucleusCallback<NucleusUser> cb) {
        async(cb, () -> NucleusUser.fromJson(get("/v1/user")));
    }

    public NucleusUser getUserSync() throws NucleusException {
        return NucleusUser.fromJson(get("/v1/user"));
    }

    public void updateUser(String firstName, String lastName, NucleusCallback<NucleusUser> cb) {
        async(cb, () -> {
            JSONObject body = new JSONObject();
            if (firstName != null) body.put("first_name", firstName);
            if (lastName != null) body.put("last_name", lastName);
            return NucleusUser.fromJson(patch("/v1/user", body));
        });
    }

    public void updatePassword(String currentPassword, String newPassword, NucleusCallback<Void> cb) {
        async(cb, () -> {
            put("/v1/user/password", new JSONObject().put("current_password", currentPassword).put("new_password", newPassword));
            return null;
        });
    }

    // --- Sessions ---
    public void refreshSession(String refreshToken, NucleusCallback<NucleusSession> cb) {
        async(cb, () -> NucleusSession.fromJson(post("/v1/sessions/refresh", new JSONObject().put("refresh_token", refreshToken))));
    }

    public NucleusSession refreshSessionSync(String refreshToken) throws NucleusException, org.json.JSONException {
        return NucleusSession.fromJson(post("/v1/sessions/refresh", new JSONObject().put("refresh_token", refreshToken)));
    }

    // --- Organizations ---
    public void getOrganizations(NucleusCallback<List<NucleusOrganization>> cb) {
        async(cb, () -> {
            JSONObject json = get("/v1/organizations");
            List<NucleusOrganization> list = new ArrayList<>();
            JSONArray arr = json.optJSONArray("organizations");
            if (arr == null && json.names() == null) return list;
            if (arr != null) { for (int i = 0; i < arr.length(); i++) list.add(NucleusOrganization.fromJson(arr.getJSONObject(i))); }
            return list;
        });
    }

    public void createOrganization(String name, String slug, NucleusCallback<NucleusOrganization> cb) {
        async(cb, () -> NucleusOrganization.fromJson(post("/v1/organizations", new JSONObject().put("name", name).put("slug", slug))));
    }

    public void getMembers(String orgId, NucleusCallback<List<NucleusMember>> cb) {
        async(cb, () -> {
            JSONObject json = get("/v1/organizations/" + orgId + "/members");
            List<NucleusMember> list = new ArrayList<>();
            JSONArray arr = json.optJSONArray("members");
            if (arr != null) { for (int i = 0; i < arr.length(); i++) list.add(NucleusMember.fromJson(arr.getJSONObject(i))); }
            return list;
        });
    }

    public void removeMember(String orgId, String memberId, NucleusCallback<Void> cb) {
        async(cb, () -> { delete("/v1/organizations/" + orgId + "/members/" + memberId); return null; });
    }

    public void createInvitation(String orgId, String email, String role, NucleusCallback<NucleusInvitation> cb) {
        async(cb, () -> NucleusInvitation.fromJson(post("/v1/organizations/" + orgId + "/invitations", new JSONObject().put("email", email).put("role", role))));
    }

    // --- Result types ---
    public static class AuthResult {
        public final NucleusUser user;
        public final NucleusSession session;
        public AuthResult(NucleusUser user, NucleusSession session) { this.user = user; this.session = session; }
    }

    public static class MfaSetupResult {
        public final String secret;
        public final String qrUri;
        public MfaSetupResult(String secret, String qrUri) { this.secret = secret; this.qrUri = qrUri; }
    }
}

File: sdks/android/src/main/kotlin/dev/nucleus/network/ApiClient.kt
package dev.nucleus.network

import androidx.credentials.Credential
import dev.nucleus.NucleusConfig
import dev.nucleus.models.NucleusOrganization
import dev.nucleus.models.NucleusSession
import dev.nucleus.models.NucleusUser
import dev.nucleus.session.SessionRestoreResponse
import dev.nucleus.session.TokenStorage
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.util.concurrent.TimeUnit

/**
 * Low-level HTTP client that communicates with the Nucleus API.
 *
 * All public methods are `suspend` so they can be called from coroutines
 * without blocking the main thread.
 */
class ApiClient internal constructor(
    private val config: NucleusConfig,
    private val tokenStorage: TokenStorage,
) {

    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
        isLenient = true
    }

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    val httpClient: OkHttpClient = OkHttpClient.Builder()
        .addInterceptor(NucleusAuthInterceptor(config, tokenStorage))
        .addInterceptor(HttpLoggingInterceptor().apply {
            level = HttpLoggingInterceptor.Level.BODY
        })
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    // ------------------------------------------------------------------
    // Auth endpoints
    // ------------------------------------------------------------------

    data class AuthResult(val session: NucleusSession, val user: NucleusUser)

    suspend fun signIn(email: String, password: String): AuthResult? {
        val body = json.encodeToString(
            kotlinx.serialization.serializer<Map<String, String>>(),
            mapOf("email" to email, "password" to password),
        )
        val request = newRequest("v1/auth/sign-in")
            .post(body.toRequestBody(jsonMediaType))
            .build()
        return executeForAuthResult(request)
    }

    suspend fun signOut() {
        val request = newRequest("v1/auth/sign-out")
            .post("{}".toRequestBody(jsonMediaType))
            .build()
        httpClient.newCall(request).execute().close()
    }

    // ------------------------------------------------------------------
    // Passkey endpoints
    // ------------------------------------------------------------------

    data class PasskeyChallenge(val challenge: String, val rpId: String, val timeout: Long)

    suspend fun getPasskeyChallenge(): PasskeyChallenge? {
        val request = newRequest("v1/auth/passkey/challenge")
            .post("{}".toRequestBody(jsonMediaType))
            .build()
        val response = httpClient.newCall(request).execute()
        if (!response.isSuccessful) return null
        val responseBody = response.body?.string() ?: return null
        val obj = json.decodeFromString<Map<String, kotlinx.serialization.json.JsonElement>>(responseBody)
        return PasskeyChallenge(
            challenge = obj["challenge"].toString().trim('"'),
            rpId = obj["rp_id"].toString().trim('"'),
            timeout = obj["timeout"].toString().trim('"').toLongOrNull() ?: 60000,
        )
    }

    suspend fun verifyPasskeyCredential(credential: Credential): AuthResult? {
        val body = json.encodeToString(
            kotlinx.serialization.serializer<Map<String, String>>(),
            mapOf("credential_type" to credential.type, "data" to credential.data.toString()),
        )
        val request = newRequest("v1/auth/passkey/verify")
            .post(body.toRequestBody(jsonMediaType))
            .build()
        return executeForAuthResult(request)
    }

    // ------------------------------------------------------------------
    // OAuth endpoints
    // ------------------------------------------------------------------

    suspend fun exchangeOAuthCode(code: String, redirectUri: String): AuthResult? {
        val body = json.encodeToString(
            kotlinx.serialization.serializer<Map<String, String>>(),
            mapOf("code" to code, "redirect_uri" to redirectUri),
        )
        val request = newRequest("v1/oauth/token")
            .post(body.toRequestBody(jsonMediaType))
            .build()
        return executeForAuthResult(request)
    }

    // ------------------------------------------------------------------
    // Session endpoints
    // ------------------------------------------------------------------

    suspend fun getSession(sessionId: String): SessionRestoreResponse? {
        val request = newRequest("v1/sessions/$sessionId").get().build()
        val response = httpClient.newCall(request).execute()
        if (!response.isSuccessful) return null
        val responseBody = response.body?.string() ?: return null
        val session = json.decodeFromString<NucleusSession>(responseBody)
        // Fetch user separately.
        val user = getUser(session.userId)
        return SessionRestoreResponse(session, user)
    }

    suspend fun refreshSession(refreshToken: String): NucleusSession? {
        val body = json.encodeToString(
            kotlinx.serialization.serializer<Map<String, String>>(),
            mapOf("refresh_token" to refreshToken),
        )
        val request = newRequest("v1/sessions/refresh")
            .post(body.toRequestBody(jsonMediaType))
            .build()
        val response = httpClient.newCall(request).execute()
        if (!response.isSuccessful) return null
        val responseBody = response.body?.string() ?: return null
        return json.decodeFromString<NucleusSession>(responseBody)
    }

    // ------------------------------------------------------------------
    // User endpoints
    // ------------------------------------------------------------------

    suspend fun getUser(userId: String): NucleusUser? {
        val request = newRequest("v1/users/$userId").get().build()
        val response = httpClient.newCall(request).execute()
        if (!response.isSuccessful) return null
        val responseBody = response.body?.string() ?: return null
        return json.decodeFromString<NucleusUser>(responseBody)
    }

    // ------------------------------------------------------------------
    // Organization endpoints
    // ------------------------------------------------------------------

    suspend fun getOrganizations(): List<NucleusOrganization> {
        val request = newRequest("v1/organizations").get().build()
        val response = httpClient.newCall(request).execute()
        if (!response.isSuccessful) return emptyList()
        val responseBody = response.body?.string() ?: return emptyList()
        return json.decodeFromString<List<NucleusOrganization>>(responseBody)
    }

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    private fun newRequest(path: String): Request.Builder {
        return Request.Builder().url("${config.baseUrl}/$path")
    }

    private fun executeForAuthResult(request: Request): AuthResult? {
        val response = httpClient.newCall(request).execute()
        if (!response.isSuccessful) return null
        val responseBody = response.body?.string() ?: return null
        // Expect { "session": { ... }, "user": { ... } }
        val element = json.parseToJsonElement(responseBody)
        val sessionJson = element.jsonObject["session"] ?: return null
        val userJson = element.jsonObject["user"] ?: return null
        val session = json.decodeFromJsonElement(NucleusSession.serializer(), sessionJson)
        val user = json.decodeFromJsonElement(NucleusUser.serializer(), userJson)
        return AuthResult(session, user)
    }

    private val kotlinx.serialization.json.JsonElement.jsonObject
        get() = this as kotlinx.serialization.json.JsonObject
}
